### PR TITLE
feat(context): allow bindings to be created from dynamic value provider classes

### PR DESCRIPTION
**Pull request by raymondfeng**
## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

File: packages/context/src/__tests__/unit/binding-inspector.unit.ts [hunk 1]
```diff
@@ -154,6 +155,72 @@ describe('createBindingFromClass()', () => {
     expect(ctx.getSync(binding.key)).to.eql('my-value');
   });
 
+  it('inspects dynamic value provider classes', () => {
+    const spec = {
+      tags: ['rest'],
+      scope: BindingScope.CONTEXT,
+    };
+
+    @bind(spec)
```
Comment (bajtos) [inline packages/context/src/__tests__/unit/binding-inspector.unit.ts:164]: Please add some documentation for LB4 app and extension developers to explain how to use `@bind` to tweak how is a dynamic value provider class bound to the context.

Comment (raymondfeng) [reply packages/context/src/__tests__/unit/binding-inspector.unit.ts:164]: Added docs.

File: packages/context/src/binding.ts [hunk 2]
```diff
@@ -267,9 +267,9 @@ function toValueFactory<T = unknown>(
  * Check if the factory is a value factory provider class
  * @param factory - A factory function or a dynamic value provider class
  */
-function isDynamicValueProviderClass(
+export function isDynamicValueProviderClass<T = unknown>(
```
Comment (deepakrkris) [inline packages/context/src/binding.ts:270]: dynamic value provider classes was already supported is it ? (not able to find any examples / docs for it)

File: packages/context/src/binding-inspector.ts [hunk 3]
```diff
@@ -85,16 +92,21 @@ export function asProvider<T>(
  * A factory function to create a template function to bind the target class
  * as a class or `Provider`.
  * @param target - Target class, which can be an implementation of `Provider`
+ * or `DynamicValueProviderClass`
  *
  * @typeParam T - Value type
  */
 export function asClassOrProvider<T>(
```
Comment (deepakrkris) [inline packages/context/src/binding-inspector.ts:99]: the focus of this PR is the change in this method , asClassOrProvider , correct ? and that is for allowing @bind() decorator and createBindingFromClass() to handle dynamic value providers is it ?

File: packages/context/src/__tests__/unit/inject.unit.ts [hunk 4]
```diff
@@ -271,3 +272,76 @@ describe('property injection', () => {
     expect(meta.foo.metadata).to.eql({x: 1, decorator: '@inject'});
   });
 });
+
+describe('inspectTargetType', () => {
+  it('handles static method injection', () => {
```
Comment (jannyHou) [inline packages/context/src/__tests__/unit/inject.unit.ts:277]: This belongs to https://github.com/strongloop/loopback-next/pull/5455 I think

File: packages/context/src/binding.ts (hunk 2 above)
Comment (raymondfeng) [reply packages/context/src/binding.ts:270]: See docs in this PR and https://github.com/strongloop/loopback-next/blob/master/docs/site/Binding.md#a-factory-function.

File: packages/context/src/binding-inspector.ts (hunk 3 above)
Comment (raymondfeng) [reply packages/context/src/binding-inspector.ts:99]: Yes.

File: packages/context/src/__tests__/unit/inject.unit.ts (hunk 4 above)
Comment (raymondfeng) [reply packages/context/src/__tests__/unit/inject.unit.ts]: Yes. just rebased.

**Pull request merged**